source/Libraries/SteamLibrary/SteamShared/SteamTagNamer.cs

```diff
@@ -46,48 +47,88 @@ private string GetPackedTagNameFilePath()
             {
                 logger.Trace($"No new tag names found for '{settings?.LanguageKey}'");
                 fileName = GetPackedTagNameFilePath();
-            }
 
-            if (!File.Exists(fileName))
-            {
-                logger.Warn($"No tag names found for language '{settings?.LanguageKey}'");
-                return new Dictionary<int, string>();
+                if (!File.Exists(fileName))
+                {
+                    logger.Warn($"No tag names found for language '{settings?.LanguageKey}'");
+                    return new Dictionary<int, string>();
+                }
             }
+            var content = GetFileContents(fileName);
+
+            return content.response.tags.ToDictionary(x => x.tagid, x => x.name);
+        }
 
-            var tagNames = Serialization.FromJsonFile<Dictionary<int, string>>(fileName);
-            return tagNames;
+        private SteamTagFile GetFileContents()
+        {
+            return GetFileContents(GetTagNameFilePath()) ?? GetFileContents(GetPackedTagNameFilePath());
         }
 
-        private static Regex tagNameRegex = new Regex(@"data-param=""tags"" data-value=""(?<id>[0-9]+)"" data-loc=""(?<name>.+?)""", RegexOptions.Compiled);
+        private SteamTagFile GetFileContents(string path)
+        {
+            if (!File.Exists(path))
+                return null;
+
+            logger.Debug("Opening " + path);
+            try
+            {
+                var content = File.ReadAllText(path, Encoding.UTF8);
+                return Serialization.FromJson<SteamTagFile>(content);
+            }
+            catch (Exception e)
+            {
+                logger.Error(e, "Failed to read " + path);
+                throw;
+            }
+        }
 
         public Dictionary<int, string> UpdateAndGetTagNames()
         {
-            string url = $"https://store.steampowered.com/search/?l={settings?.LanguageKey}";
+            var existingFile = GetFileContents();
+
+            string url = $"https://api.steampowered.com/IStoreService/GetTagList/v1?language={settings?.LanguageKey}&have_version_hash={existingFile?.response?.version_hash}";
+
             logger.Debug("Downloading " + url);
-            var content = downloader.DownloadString(url);
-            var matches = tagNameRegex.Matches(content);
+            var httpClient = new HttpClient();
```

Comment:
Why this change? You are creating new HttpClient every time this is called, which is bad (not even disposing it after use). What was wrong with the original reused downloader method?

Reply:
Because I need to check if it's a 304 result (no change, no result body). Could change it to just always download the file and only check if the content is empty. That's probably better, right?

Reply:
You are calling GET so you are downloading the body always either way. But the main point is that you should be reusing HttpClient instances, not creating new ones every time this is created.

If you want to check response code before downloading the file, you can use `HttpDownloader.GetResponseCode`, it sends HEAD request and gives you response code and headers without downloading body. Then you could just use the original downloader method.

Reply:
Removed HttpClient usage now